torch/fx/experimental/symbolic_shapes.py

```diff
@@ -4193,14 +4229,15 @@ def _create_symbolic_sizes_strides_storage_offset(
             symbolic_context,
         )
 
-        sym_sizes = [
-            self.create_symintnode(
-                sym,
-                hint=hint,
-                source=TensorPropertySource(source, TensorProperty.SIZE, i),
-            )
-            for i, (sym, hint) in enumerate(zip(size, ex_size))
-        ]
+        sym_sizes = []
+        for i, (sym, hint) in enumerate(zip(size, ex_size)):
+            node_source = TensorPropertySource(source, TensorProperty.SIZE, i)
+            node = self.create_symintnode(sym, hint=hint, source=node_source)
+            sym_sizes.append(node)
+            should_specialize = specialization and specialization.source == node_source
+            if should_specialize:
+                assert specialization  # make mypy happy
+                expect_true(specialization.check_fn(node), dont_guard=True)
```

Comment:
What's going on here? Is this that when we're creating FakeTensors and symints, we need to force them to have concrete shape and ints?

Reply:
I can think of two potential problems but not sure if you ran into them:
- to get AOTAutograd to create FakeTensors with concrete shape instead of symbolic shape, one way to do it is to trick it into pretending that the shapeenv doesn't exist: https://github.com/pytorch/pytorch/blob/main/torch/_functorch/aot_autograd.py#L491
- Dynamo shouldn't be passing the backend a symint when it's doing the specialization, it should be passing a concrete int

Reply:
actually I wonder if we can "turn dynamic shapes" off at the Dynamo level before we call the backend compiler. One way to do that could just be setting a flag on the tracing context. I'm hoping backend compilers don't need to all add an extra flag to support this feature

Reply:
So in addition to supporting specializing to a single value, I want to be able to support specializing to any expression including alignment specializations (aka x % 16 == 0, x % 8 == 0, etc).

I just refactored the code to patch the ShapeEnv before invoking the user compiler which makes all of this much cleaner. Let me know if this makes sense.